compile-time-programming/requires-expressions.md

```diff
@@ -0,0 +1,102 @@
+# Module name: Requires Expressions
+
+## Overview
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational knowledge</td>
+    <td>Define and use requires-expressions to check satisfaction of expressions by given parameters</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td>Define and use requires-expressions to check properties of expressions</td>
+  </tr>
+  <tr>
+    <td>Further studies</td>
+    <td></td>
+  </tr>
+</table>
+
+## Motivation
+
+Requires-expressions allow a developer to perform compile-time evaluation on the validity of other expressions. These are fundamental to the ability to write concepts.[[Compile-time programming: concepts]][1]
+
+## Topic introduction
+
+Requires-expressions are compile-time predicates which evaluate to true when their specified set of expressions are all valid for a given set of inputs.
+
+## Foundational knowledge: Writing requires-expressions
+
+### Background/Required Knowledge
+
+A student is able to:
+
+* Write and use a function template [[Compile-time programming: function templates]][2]
+* Differentiate between expressions and statements
+
+It is helpful if:
+
+* A student is aware of the constraints imposed by their template definitions
```

Comment:
Ahhh, hmmm... was going for something like

be aware that:
```
template<typename T>
void foo(T t) {
    std::cout << t;
}
```
Has requirements that `T` be copyable and that there be some overload matching `operator<<(std::ostream&, T)`

Reply:
Is there an alternative wording you can suggest that might help?

Reply:
I think it'd be good to avoid the word "constraints" here, as it (now) has a technical meaning. What you're talking about is the informal _structural requirements(?)_ imposed by the bodies of even unconstrained function templates.

I'm not very happy with my phrase "structural requirements." I'm happy with "requirements" over "constraints," but, without an adjective, "requirements" alone is too easy to mistake for something related to `requires`. (A `requires`-clause imposes a constraint, not a requirement: ick.)

Reply:
would `A student is aware of the syntactic requirements, both direct and indirect, imposed by their usage of the type within their template.` sound better?

Reply:
How about `The student is aware that templates effectively impose syntactic and semantic requirements on their type parameters based on how those parameters are used`; plus `The student is able to give practical examples of syntactic requirements and semantic requirements`?
(Although at some point we slip from "it's nice if the student arrives knowing this" to "this is actually part of the thing we're trying to teach" :))

Reply:
I was leaning more into the "It would be nice for the student to know..." where that knowledge creates a simple baseline for understanding the motivation for having such checks. Maybe something more like "The student is aware that attempts to specialize the template with types or values which do not match otherwise unstated assumptions will cause errors within the template... " something like that... need to wake the brain up before I can actually do the proper words...
The point being, perhaps we want to convey the student would benefit from having had compiler errors for trying to use types that dont work with their template definition.

Reply:
`The student is aware that attempting to specialize the template with types or values which do not match otherwise unstated assumptions will cause errors within the template.` :+1: I like that.

Reply:
change applied